File: DotNetGotchas/VBDN/RTTI/Modified1/RTTIExample/Trainer.vb
'Trainer.vb
Public Class Trainer
    Public Sub Train(ByVal anAnimal As Animal)
        anAnimal.Eat()

        'Using RTTI

        If TypeOf anAnimal Is Dog Then
            Dim doggie As Dog = CType(anAnimal, Dog)
            doggie.Bark()
        ElseIf TypeOf anAnimal Is Cat Then
            Dim aCat As Cat = CType(anAnimal, Cat)
            aCat.Meow()
        End If
    End Sub
End Class

File: DotNetGotchas/CSharp/RTTI/Modified1/RTTIExample/Trainer.cs
//Trainer.cs
using System;

namespace RTTIExample
{
	public class Trainer
	{
		public void Train(Animal anAnimal)
		{
			anAnimal.Eat();

			//Using RTTI

			if (anAnimal is Dog)
			{
				Dog doggie = (Dog) anAnimal;
				doggie.Bark();
			}
			else if (anAnimal is Cat)
			{
					Cat aCat = (Cat) anAnimal;
					aCat.Meow();
			}

//			Dog doggie = anAnimal as Dog;
//			if (doggie != null)
//			{
//				doggie.Bark();
//			}
		}
	}
}
